src/hardware/bk-1856d/protocol.c

```diff
@@ -17,25 +17,344 @@
+	serial = sdi->conn;
+	if (!devc || !serial)
+		return;
+
+	if (devc->gate_time > 3)
+		devc->gate_time = 3;
+	cfg = &(gate_time_config_commands[devc->gate_time]);
+
+	sr_info("%s", cfg->info);
+
+	if (serial_write_blocking(serial, cfg->cmd, LENGHT_OF_CMD,
+			serial_timeout(serial, LENGHT_OF_CMD)) < 1) {
+		sr_err("unable to send gate time command");
+	}
+	g_usleep(cfg->sleep_time);
+}
+
+static void bk_1856d_request_data(const struct sr_dev_inst *sdi)
+{
+	struct dev_context *devc;
+	struct sr_serial_dev_inst *serial;
+
+	if (!sdi)
+		return;
+
+	devc = sdi->priv;
+	serial = sdi->conn;
+	if (!devc || !serial)
+		return;
+
+	sr_spew("requesting data");
+
+	if (serial_write_blocking(serial, DATA_REQ, LENGHT_OF_CMD,
+			serial_timeout(serial, LENGHT_OF_CMD)) < 1) {
+		sr_err("unable to send request data command");
+	}
+}
+
+SR_PRIV void bk_1856d_init(const struct sr_dev_inst *sdi)
+{
+	struct dev_context *devc;
+	struct sr_serial_dev_inst *serial;
+
+	if (!sdi)
+		return;
+
+	devc = sdi->priv;
+	serial = sdi->conn;
+	if (!devc || !serial)
+		return;
+
+	devc->buffer_level = 0;
+	sr_sw_limits_acquisition_start(&(devc->sw_limits));
+	serial_flush(serial);
+
+	bk_1856d_send_input_sel(sdi);
+
+	bk_1856d_send_gate_time(sdi);
+	bk_1856d_request_data(sdi);
+}
+
+static int bk_1856d_check_for_zero_message(struct dev_context *devc)
+{
+	int zero_found;
+	int has_data;
+
+	zero_found = 0;
+	has_data = 0;
+
+	for (int idx = 0; idx < BK1856D_MSG_SIZE - 1; ++idx) {
+		if (devc->buffer[idx] == ' ') {
+			continue;
+		}
+		else if (devc->buffer[idx] == '0' && zero_found == 0) {
+			zero_found = 1;
+			continue;
+		}
+		else {
+			has_data = 1;
+			break;
+		}
+	}
+	return !has_data;
+}
+
+static void bk_1856d_send_packet(
+	const struct sr_dev_inst *sdi, double freq_value, int digits)
+{
+	struct sr_datafeed_packet packet;
+	struct sr_datafeed_analog analog;
+	struct sr_analog_encoding encoding;
+	struct sr_analog_meaning meaning;
+	struct sr_analog_spec spec;
+
+	sr_analog_init(&analog, &encoding, &meaning, &spec, digits);
+	analog.meaning->mq = SR_MQ_FREQUENCY;
+	analog.meaning->unit = SR_UNIT_HERTZ;
+	analog.meaning->channels = sdi->channels;
+	analog.num_samples = 1;
+	analog.data = &freq_value;
+	analog.encoding->unitsize = sizeof(freq_value);
+	analog.encoding->is_float = TRUE;
+	analog.encoding->digits = digits;
+
+	packet.type = SR_DF_ANALOG;
+	packet.payload = &analog;
+	sr_session_send(sdi, &packet);
+}
+
+static void bk_1856d_parse_message(const struct sr_dev_inst *sdi)
+{
+	struct dev_context *devc;
+	struct sr_serial_dev_inst *serial;
+
+	double freq_value;
+	int digits;
+	char *endPtr, *dotPtr;
+
+	if (!sdi->priv || !sdi->conn)
+		return;
+
+	devc = sdi->priv;
+	serial = sdi->conn;
+
+	/* check for cr at end of message */
+	if (devc->buffer[BK1856D_MSG_SIZE-1] != '\xD') {
+		sr_err("expected cr at end of message.");
+		devc->buffer_level = 0;
+		serial_flush(serial);
+		bk_1856d_send_input_sel(sdi);
+		bk_1856d_send_gate_time(sdi);
```

Comment:
You seem to select input and send gate time every time you interact with the device. Is that actually necessary? Do things break when you leave these out?